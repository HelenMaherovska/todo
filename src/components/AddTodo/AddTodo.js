import React, { useState } from "react";
import PropTypes from "prop-types";

import { connect } from "react-redux";

import { addTodo } from "actions/todoActions";
import "components/AddTodo/AddTodo.css";

export const AddTodo = props => {
  const [inputValue, setInputValue] = useState("");

  const onAddTodo = () => {
    const { addTodo } = props;

    setInputValue("");

    return inputValue.trim() ? addTodo(inputValue) : null;
  };

  return (
    <div className="add-todo-wrapper input-group d-flex pr-0 pl-0 justify-content-between">
      <input
        type="text"
        className="form-control mr-4"
        value={inputValue}
        onChange={e => setInputValue(e.target.value)}
      />
      <button type="submit" className="btn btn-success" onClick={onAddTodo}>
        Add Todo
      </button>
    </div>
  );
};

AddTodo.propTypes = {
  addTodo: PropTypes.func.isRequired
};

const mapDispatchToProps = {
  addTodo
};

export default connect(
  null,
  mapDispatchToProps
)(AddTodo);
