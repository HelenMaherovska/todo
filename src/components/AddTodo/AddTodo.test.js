import React from "react";
import { render, fireEvent } from "react-testing-library";
import "jest-dom/extend-expect";
import { AddTodo } from "components/AddTodo/AddTodo";

describe("AddTodo", () => {
  const text = "Learn unit testing";
  const mockAddTodo = jest.fn();

  const { container } = render(<AddTodo addTodo={mockAddTodo} />);

  it("renders correctly", () => {
    expect(container).toMatchSnapshot();
  });

  describe("Initial state", () => {
    it("should have the `Add Todo` button", () => {
      expect(container.querySelector(".btn-success").textContent).toBe(
        "Add Todo"
      );
    });
  });

  describe("before typing into the add todo input", () => {
    const inputVal = container.querySelector("input");

    fireEvent.change(inputVal, { target: { value: text } });

    it("updates the inputValue state", () => {
      expect(container.querySelector("input")).toHaveAttribute("value", text);
    });
  });

  describe("when clicking the `Add Todo` button", () => {
    beforeEach(() => {
      const btn = container.querySelector(".btn-success");
      fireEvent.click(btn);
    });

    it("successfully calls the onClick handler", () => {
      expect(mockAddTodo.mock.calls.length).toEqual(1);
    });

    it("dispatches the `addTodo()` it receives from props", () => {
      expect(mockAddTodo).toHaveBeenCalledWith(text);
    });
  });

  describe("after clicking the `Add Todo` button", () => {
    it("inputValue in state should be empty", () => {
      expect(container.querySelector("input")).toHaveAttribute("value", "");
    });
  });
});
