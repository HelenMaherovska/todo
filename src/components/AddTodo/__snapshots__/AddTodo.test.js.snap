// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodo renders correctly 1`] = `
<div>
  <div
    class="add-todo-wrapper input-group d-flex pr-0 pl-0 justify-content-between"
  >
    <input
      class="form-control mr-4"
      type="text"
      value="Learn unit testing"
    />
    <button
      class="btn btn-success"
      type="submit"
    >
      Add Todo
    </button>
  </div>
</div>
`;
